src/load.cpp

```diff
@@ -334,12 +352,29 @@ void LoadOutput::control()
             (LOAD_CURRENT_MAX * LOAD_CURRENT_MAX) * (MOSFET_MAX_JUNCTION_TEMP - 25)
         ) / (MOSFET_THERMAL_TIME_CONSTANT * CONTROL_FREQUENCY);
 
-    if (junction_temperature > MOSFET_MAX_JUNCTION_TEMP) {
+    if (junction_temperature > MOSFET_MAX_JUNCTION_TEMP
+        || port->current > LOAD_CURRENT_MAX * 2)
+    {
         switch_set(false);
         usb_set(false);
+        leds_flicker(LED_LOAD);
+        state = LOAD_STATE_OFF_OVERCURRENT;
+        log_data.error_flags |= (1 << ERR_LOAD_OVERCURRENT);
+        print_error("Load overcurrent detected\n");
+        overcurrent_timestamp = time(NULL);
+    }
+
+    // overcurrent detected because of voltage dip by 25%
+    if (port->voltage < voltage_prev * 0.75 && voltage_prev != 0) {
+        switch_set(false);
+        usb_set(false);
+        leds_flicker(LED_LOAD);
         state = LOAD_STATE_OFF_OVERCURRENT;
+        log_data.error_flags |= (1 << ERR_LOAD_VOLTAGE_DIP);
+        print_error("Load voltage dip detected\n");
         overcurrent_timestamp = time(NULL);
     }
+    voltage_prev = port->voltage;
 
     static int debounce_counter = 0;
```

Comment:
No, because then they are the same when we do the if clause, so it will never become true.

Reply:
Sorry, I think I did not get this. I was thinking if that is fine to move the below line to the top of the function,
static int debounce_counter = 0;

Reply:
Ahh, sorry, my mistake, I thought you were talking about `voltage_prev`, as that was highlighted by the diff view.

I initialized the debounce counter here because it is used by the block that follows the initialization. For sure, one could argue that static variables should all be declared at the beginning of a function. Is that typically done? Don't know.